eng/tools/dependency-testing/index.js

```diff
@@ -0,0 +1,362 @@
+      type: "string",
+      describe:
+        "name of the artifact to be incremented (e.g. azure-keyvault-secrets), will be translated to @azure/(package) format",
+      demandOption: true
+    },
+    "repo-root": {
+      type: "string",
+      default: "../../../",
+      describe: "root of the repository (e.g. ../../../)",
+      demandOption: true
+    },
+    "version-type": {
+      type: "string",
+      describe: "whether you want to test max or min or same version of dependencies",
+      demandOption: true
+    },
+    "source-dir": {
+      type: "string",
+      describe: "complete local path of the directory of the repo",
+      demandOption: true
+    },
+    "test-folder": {
+      type: "string",
+      default: "test",
+      describe: "whether to point at test or test/public",
+      demandOption: false
+    },
+    "dry-run": {
+      type: "boolean"
+    }
+  })
+  .help().argv;
+
+const fs = require("fs");
+const path = require("path");
+const semver = require("semver");
+const packageUtils = require("eng-package-utils");
+// crossSpawn is used because of its ability to better handle corner cases that break when using spawn in Windows environments.
+// For more details see - https://www.npmjs.com/package/cross-spawn
+var crossSpawn = require("cross-spawn");
+
+function outputTestPath(projectFolderPath, sourceDir, testFolder) {
+  const projectPath = path.join(sourceDir, projectFolderPath);
+  const testPath = path.join(projectPath, testFolder);
+  console.log(`##vso[task.setvariable variable=PackageTestPath]${testPath}`)
+  console.log(`Emitted variable "PackageTestPath" with content: ${testPath}`)
+}
+
+async function insertPackageJson(repoRoot, packageJsonContents, targetPackagePath, targetPackageName, versionType, testFolder) {
+  const testPath = path.join(targetPackagePath, testFolder);
+  var templateJson = await packageUtils.readFileJson("./templates/package.json");
+  //console.log(templateJson);
+  var testPackageJson = templateJson;
+  testPackageJson.name = packageJsonContents.name.replace("@azure/", "azure-") + "-test";
+  testPackageJson.devDependencies = {};
+  depList = {};
+  var projectFolder = path.basename(targetPackagePath);
+  var projectDir = path.basename(path.dirname(targetPackagePath));
+  // console.log(projectFolder);
+  // console.log(projectDir);
+  //depList[targetPackageName] = "..";//works
+  //depList[targetPackageName] = "../../../" + projectDir + "/" + projectFolder;
+  //depList[targetPackageName] = "..";
+  var allowedVersionList = {};
+  depList[targetPackageName] = packageJsonContents.version;//works
+  allowedVersionList[targetPackageName] = depList[targetPackageName];
+  for (const package of Object.keys(packageJsonContents.dependencies)) {
+    depList[package] = packageJsonContents.dependencies[package];
+    if (package.startsWith("@azure/")) {
+      depList[package] = await findAppropriateVersion(package, packageJsonContents.dependencies[package], repoRoot, versionType);
```

Comment:
So if these packages are in the rush workspace and we are using rush to install dependencies does it complain when you reference another version then the one in the workspace?